src/Bedrock.Framework.Experimental/Protocols/Kafka/KafkaClientConnectionContext.cs

```diff
@@ -0,0 +1,131 @@
+using System.Diagnostics;
+using System.Collections.Concurrent;
+using System;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Bedrock.Framework.Protocols.Kafka.Messaging;
+using Bedrock.Framework.Protocols.Kafka.Utils;
+
+namespace Bedrock.Framework.Protocols.Kafka
+{
+    /// <summary>
+    /// Encapsulates a client Kafka connection that channels requests to an underlying connection
+    /// in the order that they were written; and reads responses in the same order.
+    /// </summary>
+    public class KafkaClientConnectionContext : IKafkaClientConnectionContext
+    {
+        private readonly Channel<(object, IKafkaMessageDescriptor, TaskCompletionSource<object>)> _requestChannel =
+            Channel.CreateUnbounded<(object, IKafkaMessageDescriptor, TaskCompletionSource<object>)>(
+                new UnboundedChannelOptions()
+                {
+                    SingleReader = true,
+                    SingleWriter = false
+                });
+
+        private readonly ConcurrentQueue<(int, IKafkaMessageDescriptor, TaskCompletionSource<object>)> _requestQueue =
+            new ConcurrentQueue<(int, IKafkaMessageDescriptor, TaskCompletionSource<object>)>();
+
+        private readonly IMessageWriter<KafkaRequest> _requestWriter;
+        private readonly IMessageReader<KafkaResponse> _responseReader;
+
+        private int _correlationId = 0;
+
+        /// <summary>
+        /// Initializes a new <see cref="KafkaClientConnectionContext"/>.
+        /// </summary>
+        /// <param name="requestWriter">An <see cref="IMessageWriter{KafkaRequest}"/> that can write Kafka requests to
+        /// a buffer.</param>
+        /// <param name="responseReader">An <see cref="IMessageReader{KafkaResponse}"/> that can read Kafka responses from
+        /// a buffer.</param>
+        public KafkaClientConnectionContext(IMessageWriter<KafkaRequest> requestWriter, IMessageReader<KafkaResponse> responseReader)
+        {
+            _requestWriter = requestWriter;
+            _responseReader = responseReader;
+        }
+
+        /// <inheritdoc/>
+        public Task StartAsync(ConnectionContext connectionContext, CancellationToken cancellationToken = default)
+        {
+            _ = WriteRequestsAsync(connectionContext, cancellationToken);
+            _ = ReadResponsesAsync(connectionContext, cancellationToken);
+
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc/>
+        public async Task<TResponseSchema> SendAsync<TRequestSchema, TResponseSchema>(
+            TRequestSchema request,
+            IKafkaMessageDescriptor messageDescriptor,
+            CancellationToken cancellationToken = default)
+        {
+            var tcs = new TaskCompletionSource<object>();
+
+            _requestChannel.Writer.TryWrite((request, messageDescriptor, tcs));
```

Comment:
maybe want a BoundedChannel and use `WaitAsync` here?
Gives the client a little "back pressure"?
librdkafka has various config to limit "queue" sizes.
Just a thought

Reply:
Yes. I expect some of this to come out in the wash!